controllers/applicationvolumereplication_controller_test.go

```diff
@@ -133,413 +139,524 @@ func (s *FakeS3StoreWrapper) DownloadPVs(ctx context.Context, r client.Reader,
+
+	vrgClusterRoleBinding := &rbacv1.ClusterRoleBinding{}
+
+	err = yaml.Unmarshal(vrgClusterRoleManifest.RawExtension.Raw, &vrgClusterRoleBinding)
+	Expect(err).NotTo(HaveOccurred())
+
+	manifestLookupKey := types.NamespacedName{
+		Name:      controllers.BuildManifestWorkName(subscription.Name, subscription.Namespace, "vrg"),
+		Namespace: managedCluster,
+	}
+	createdManifest := &ocmworkv1.ManifestWork{}
+
+	Eventually(func() bool {
+		err := k8sClient.Get(context.TODO(), manifestLookupKey, createdManifest)
+
+		return err == nil
+	}, timeout, interval).Should(BeTrue())
+
+	// 5.1 verify that VRG CR has been created and added to the ManifestWork
+	Expect(len(createdManifest.Spec.Workload.Manifests)).To(Equal(1))
+	vrgClientManifest := createdManifest.Spec.Workload.Manifests[0]
+
+	Expect(vrgClientManifest).ToNot(BeNil())
+
+	vrg := &ramendrv1alpha1.VolumeReplicationGroup{}
+
+	err = yaml.Unmarshal(vrgClientManifest.RawExtension.Raw, &vrg)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(vrg.Name).Should(Equal(subscription.Name))
+	Expect(vrg.Spec.PVCSelector.MatchLabels["appclass"]).Should(Equal("gold"))
+}
+
+func getManifestWorkCount(homeClusterNamespace string) int {
+	manifestWorkList := &ocmworkv1.ManifestWorkList{}
+	listOptions := &client.ListOptions{Namespace: homeClusterNamespace}
+
+	Expect(k8sClient.List(context.TODO(), manifestWorkList, listOptions)).NotTo(HaveOccurred())
+
+	return len(manifestWorkList.Items)
+}
+
+func verifyAVRStatusExpectation(subscription *subv1.Subscription, homeCluster, peerCluster, prevHomeCluster string) {
+	avrLookupKey := types.NamespacedName{
+		Name:      ApplicationVolumeReplicationName,
+		Namespace: ApplicationVolumeReplicationNamespaceName,
+	}
+
+	updatedAVR := &ramendrv1alpha1.ApplicationVolumeReplication{}
+
+	Eventually(func() bool {
+		err := k8sClient.Get(context.TODO(), avrLookupKey, updatedAVR)
+
+		if d, found := updatedAVR.Status.Decisions[subscription.Name]; err == nil && found {
+			return d.HomeCluster == homeCluster
+		}
+
+		return false
+	}, timeout, interval).Should(BeTrue())
+
+	// 7.0 check that the home and peer clusters have been selected.
+	Expect(updatedAVR.Status.Decisions[subscription.Name].HomeCluster).Should(Equal(homeCluster))
+	Expect(updatedAVR.Status.Decisions[subscription.Name].PeerCluster).Should(Equal(peerCluster))
+	Expect(updatedAVR.Status.Decisions[subscription.Name].PrevHomeCluster).Should(Equal(prevHomeCluster))
+}
+
+func expectSubscriptionIsPaused(subscription *subv1.Subscription) {
+	subLookupKey := types.NamespacedName{Name: subscription.Name, Namespace: subscription.Namespace}
+	updatedSubscription := &subv1.Subscription{}
+	err := k8sClient.Get(context.TODO(), subLookupKey, updatedSubscription)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(updatedSubscription.GetLabels()).ToNot(BeNil())
+	Expect(updatedSubscription.GetLabels()[subv1.LabelSubscriptionPause]).Should(Equal("true"))
+}
+
+func waitForCompletion() {
+	Eventually(func() bool {
```

Comment:
I have a PR ready to go out that adds the lastKnownState to the avr.status, which is per subscription.  But at this moment it is not very useful because it does not affect the operation of the AVR.  It is for informational only.  Having said that; there are definitely other ways of solving this problem.  I could definitely refactor it later and remove the callback if I find it not useful beyond the unit test.